### ISSUE TITLE
XMLParser doesn't conform to the XML specification for elements

### ISSUE THREAD

**Issue by MewPurPur** [godotengine/godot#89668]
### Tested versions

Reproducible in 4.2.1

### System information

Godot v4.2.1.stable - Pop!_OS 22.04 LTS - X11 - Vulkan (Forward+) - integrated Intel(R) Graphics (ADL GT2) () - 12th Gen Intel(R) Core(TM) i5-1235U (12 Threads)

### Issue description

xml_parser.cpp rarely changes, so this issue likely existed in Godot's entire lifetime.

The XML specification defines three types of tags: Start-tags e.g <svg>, End-tags e.g. </svg>, and Empty-Element Tags e.g. <svg/>. The particular syntax is as follows:

**Attributes section:** You have multiple attributes that look like `AttributeName="AttributeValue"`, with a possible arbitrary amounts of whitespace around the equal sign (space, \t, \n, \r) and at the start (at least one is required at the start).

**Tag name:** Can be a bunch of characters, but whitespaces cannot be part of it. This implies that a whitespace should terminate the tag name.

**Start-tag:** `<`, followed by the tag name, followed by an attribute section, followed by an arbitrary amount of whitespace (not required), followed by `>`.

**End-tag:** `</`, followed by the tag name, followed by an arbitrary amount of whitespace (not required), followed by `>`.

**Tag for empty elements:** Exactly like a start-tag, but ends in `/>` instead of `>`.

Relevant section of the specification: https://www.w3.org/TR/REC-xml/#sec-starttags

-----

XMLParser parses certain non-conforming structures, of which some are for sure unintentional. Namely:

- Doesn't seem to consider the validity of tag/attribute names at all.
- Doesn't wait for the final ">" in order to consider something a tag.
- Worst of all and 100% a bug: Whitespace of end-tags gets considered as part of the tag name.

### Steps to reproduce

```gdscript
func _ready():
	var parser = XMLParser.new()
	
	# Example 1: A non-conforming start-tag treated as an element.
	parser.open_buffer("<svg".to_utf8_buffer())
	while parser.read() == OK:
		if parser.get_node_type() == XMLParser.NODE_ELEMENT:
			print("Element")
	
	# Example 2: A non-conforming end-tag treated as an element end.
	parser.open_buffer("</".to_utf8_buffer())
	while parser.read() == OK:
		if parser.get_node_type() == XMLParser.NODE_ELEMENT_END:
			print("Element end")
	
	# Example 3: A non-conforming empty-element tag treated as an element.
	parser.open_buffer("<svg/".to_utf8_buffer())
	while parser.read() == OK:
		if parser.get_node_type() == XMLParser.NODE_ELEMENT:
			print("Element")
	
	# Example 4: Whitespace for end-tags becoming part of the tag name.
	parser.open_buffer("</svg   >".to_utf8_buffer())
	while parser.read() == OK:
		if parser.get_node_type() == XMLParser.NODE_ELEMENT_END:
			prints(parser.get_node_name() == "svg   ")
	
	# Example 5: Invalid tag name.
	parser.open_buffer("<42>".to_utf8_buffer())
	while parser.read() == OK:
		if parser.get_node_type() == XMLParser.NODE_ELEMENT:
			prints("Element")
```

### Minimal reproduction project (MRP)

N/A just copy the code I provided into a node script.

**Comment by KrukDallan** [comment]
The problems are in the "XMLParser::_parse_opening_xml_element()" and "XMLParser::_parse_closing_xml_element()" methods. I'm looking into it, I'll let you know when I'll open the PR

**Comment by MewPurPur** [comment]
Which of these do you believe are issues? I'm realizing now that XMLParser doesn't exactly try to detect syntax errors, but the last one is definitely a problem even then.

**Comment by KrukDallan** [comment]
> Which of these do you believe are issues? I'm realizing now that XMLParser doesn't exactly try to detect syntax errors, but the last one is definitely a problem even then.

You are right, the XMLParser does not check for syntax errors, as well as for other forbidden cases (i.e. a tag starting with 'xml'). I have added some of these checks, and I'm trying to open a pull request, but since it's my first time doing it it's taking a while, sorry about that

**Comment by KrukDallan** [comment]
I opened the PR, but as I said previously I didn't cover every problem, mainly because I'm not 100% sure I'm in line with their way of doing things, so I have to wait for a feedback. Nonetheless, in the next days I'll continue working on it, and if necessary I'll apply the necessary changes.

Have a nice day!

**Comment by KrukDallan** [comment]
> Would it be worth it to replace the custom XML parser with an open source library? There are a number to choose from. I'd expect them to be better tested and their interfaces are easier to use. Or is there some advantage to the custom parser?

I agree, there are many open source xml parsers to choose from, hence there would be no need to reinvent the wheel

**Comment by MewPurPur** [comment]
I'm not sure if anything other than the "Whitespace of end-tags gets considered as part of the tag name." part is objectively a bug. XMLParser doesn't have error reporting, and it considers other types of nodes preemptively too. Normally, comments look like `<!-- A comment -->`, but as soon as XMLParser sees "<!" it says there's a comment node. CDATA looks like `<![]>` at the minimum, usually `<![CDATA[]]>`, but XMLParser only needs to see "<![" to say there's a CDATA node.

**Comment by KrukDallan** [comment]
I'm not sure if "bug" is the correct term, however the examples you reported should raise exceptions standing to the xml specification. 

If you want, you can try to build godot from my fork, in particular the one in the "xml" branch, as I updated the XMLParser object. Now it should be more compliant to the xml specification. Let me know if you try!
